src/params/types.ts

```diff
@@ -28,6 +30,19 @@
 export abstract class Expression<T extends string | number | boolean | string[]> {
   // Returns the Expression's runtime value, based on the CLI's resolution of params.
   value(): T {
+    if (process.env.FIREBASE_DISCOVERY_DIR) {
+      logger.warn(
+        `Parametrized expression ${this.toString()}.value() invoked during Function discovery. This is usually a mistake.`
+      );
+      logger.warn(
+        `To configure a Function with an expression object, use it directly without calling .value().`
```

Comment:
```suggestion
        `To configure a function with a parameter, use it directly without calling .value().`
```

Some suggestions:

1) Let's avoid using unfamiliar Expression, even if it's technically more correct term.

2) If we can make this message super verbose, can we add an example?